x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/CommandLineHttpClient.java

```diff
@@ -253,4 +304,48 @@ public void checkClusterHealthWithRetriesWaitingForCluster(String username, Secu
     public static URL createURL(URL url, String path, String query) throws MalformedURLException, URISyntaxException {
         return new URL(url, (url.toURI().getPath() + path).replaceAll("/+", "/") + query);
     }
+
+    public static String apiKeyHeaderValue(SecureString apiKey) {
+        CharBuffer chars = CharBuffer.allocate(apiKey.length());
+        byte[] charBytes = null;
+        try {
+            chars.put(apiKey.getChars());
+            charBytes = CharArrays.toUtf8Bytes(chars.array());
+
+            //TODO we still have passwords in Strings in headers. Maybe we can look into using a CharSequence?
+            String apiKeyToken = Base64.getEncoder().encodeToString(charBytes);
+            return "ApiKey " + apiKeyToken;
+        } finally {
+            Arrays.fill(chars.array(), (char) 0);
+            if (charBytes != null) {
+                Arrays.fill(charBytes, (byte) 0);
+            }
+        }
+    }
+
+    /**
+     * Returns a TrustManager to be used in a client SSLContext, which trusts all certificates that are signed
+     * by a specific CA certificate ( identified by its SHA256 fingerprint, {@code pinnedCaCertFingerPrint} )
+     */
+    private TrustManager fingerprintTrustingTrustManager(String pinnedCaCertFingerprint) {
```

Comment:
It is intentional that we do pinning to the CA certificate instead of the leaf certificate. The Enroll Kibana API could* return multiple addresses for different existing nodes in the cluster and we wanted a way to be able to bootstrap trust for a connection to any of these nodes, with a singe piece of information.


\* It currently doesn't, it returns addresses from the node that serves the API request, but the original design did and future potential work could introduce this.

Reply:
I know we only intend to support our tooling here, so this isn't really an issue, but I think I would prefer to pin to `chain[1]` rather than `chain[length-1]`
That is, pin to the issuer, not the root CA.

However, that then fails for self-signed certs, so there's no perfect solution that's also concise.